File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/date/GetStartOfTimeStamp.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.function.reactive.AbstractReactiveFunction;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.FunctionSignature;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;

import reactor.core.publisher.Mono;

public class GetStartOfTimeStamp extends AbstractReactiveFunction {

    private static final String DATE = "isodate";

    private static final String TIME_UNIT = "unit";

    @Override
    public FunctionSignature getSignature() {
        return new FunctionSignature().setName("GetStartOfTimeStamp")
                .setNamespace(Namespaces.DATE)
                .setParameters(Map.ofEntries(
                        Parameter.ofEntry(DATE, Schema.ofRef(Namespaces.DATE + ".timeStamp")),
                        Parameter.ofEntry(TIME_UNIT, Schema.ofString(TIME_UNIT).setEnums(
                                List.of()))))
                .setEvents(Map.of());
    }

    @Override
    protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {
        return null;
    }

}

File: kirun-js/src/engine/function/system/date/GetStartOfTimestamp.ts
import { KIRuntimeException } from '../../../exception/KIRuntimeException';
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AdjustTimestamp } from '../../../util/date/AdjustTimestampUtil';
import isValidZuluDate from '../../../util/date/isValidISODate';
import { AbstractFunction } from '../../AbstractFunction';

const iso8601Pattern =
    /^([+-]?\d{6}|\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[1-2]\d|3[0-1])T([0-1]\d|2[0-3]):([0-5]\d):([0-5]\d)(\.\d{3})?(Z|([+-]([01]\d|2[0-3]):([0-5]\d)))?$/;

const VALUE = 'isodate';
const OUTPUT = 'result';
const UNIT = 'unit';
const SIGNATURE = new FunctionSignature('getStartOfTimestamp')
    .setNamespace(Namespaces.DATE)
    .setParameters(
        new Map([
            [VALUE, new Parameter(VALUE, Schema.ofRef(`${Namespaces.DATE}.timeStamp`))],
            [
                UNIT,
                new Parameter(
                    UNIT,
                    Schema.ofString(UNIT).setEnums([
                        'year',
                        'month',
                        'week',
                        'quarter',
                        'date',
                        'hour',
                        'minute',
                        'second',
                    ]),
                ),
            ],
        ]),
    )
    .setEvents(new Map([Event.outputEventMapEntry(new Map([[OUTPUT, Schema.ofString(OUTPUT)]]))]));

export class GetStartOfTimestamp extends AbstractFunction {
    public getSignature(): FunctionSignature {
        return SIGNATURE;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let date: string = context.getArguments()?.get(VALUE);

        let field: string = context.getArguments()?.get(UNIT);

        if (!isValidZuluDate(date)) throw new KIRuntimeException(`Invalid ISO 8601 Date format.`);

        const adjustTimestamp = new AdjustTimestamp();

        return new FunctionOutput([
            EventResult.outputOf(
                new Map([[OUTPUT, adjustTimestamp.getStartWithGivenField(date, field)]]),
            ),
        ]);
    }
}
